File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {

        char[] sarr=s.toCharArray();
        int longest=0;
        int l=0;

        Set<Character> hs= new HashSet<>();
        for (int r=0;r<sarr.length;r++)
        {
            while(hs.contains(sarr[r])){
                hs.remove(sarr[l]);
                l++;
            }
            hs.add(sarr[r]);
            longest=Math.max(longest,r-l+1);

        }
        return longest;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        l=0
        r=0

        cmap={}
        m=0

        while(r<len(s)):
            if s[r] in cmap:
                l = max(cmap[s[r]] + 1, l)
            cmap[s[r]]=r
            m = max(m,r-l+1)
            r+=1
        
        return m
